src/perf/lib/ThroughputClient.cpp

```diff
@@ -0,0 +1,410 @@
+
+--*/
+
+#ifdef QUIC_CLOG
+#include "ThroughputClient.cpp.clog.h"
+#endif
+
+#ifndef _KERNEL_MODE
+#define QUIC_TEST_APIS 1
+#endif
+#include "msquichelper.h"
+#include "ThroughputClient.h"
+#include "ThroughputCommon.h"
+
+static uint8_t* RawIoBuffer{nullptr};
+
+ThroughputClient::ThroughputClient(
+    ) {
+    QuicZeroMemory(&LocalIpAddr, sizeof(LocalIpAddr));
+    if (Session.IsValid()) {
+        Session.SetAutoCleanup();
+    }
+}
+
+QUIC_STATUS
+ThroughputClient::Init(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[]
+    ) {
+    Port = THROUGHPUT_DEFAULT_PORT;
+    TryGetValue(argc, argv, "port", &Port);
+
+    const char* Target;
+    if (!TryGetValue(argc, argv, "target", &Target)) {
+        WriteOutput("Must specify '-target' argument!\n");
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    uint16_t ip;
+    if (TryGetValue(argc, argv, "ip", &ip)) {
+        switch (ip) {
+        case 4: RemoteFamily = AF_INET; break;
+        case 6: RemoteFamily = AF_INET6; break;
+        }
+    }
+
+    TryGetValue(argc, argv, "length", &Length);
+    if (Length == 0) {
+        WriteOutput("Must specify a positive 'length'\n");
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    const char* localAddress;
+    if (TryGetValue(argc, argv, "bind", &localAddress)) {
+        if (!ConvertArgToAddress(localAddress, 0, &LocalIpAddr)) {
+            WriteOutput("Failed to decode IP address: '%s'!\nMust be *, a IPv4 or a IPv6 address.\n", localAddress);
+            return QUIC_STATUS_INVALID_PARAMETER;
+        }
+    }
+
+    // TODO: Core, since we need to support kernel mode
+    #if _WIN32 && !_KERNEL_MODE
+    uint16_t compartmentid;
+    if (TryGetValue(argc, argv, "comp",  &compartmentid)) {
+        NETIO_STATUS status;
+        if (!NETIO_SUCCESS(status = SetCurrentThreadCompartmentId(compartmentid))) {
+            WriteOutput("Failed to set compartment ID = %d: 0x%x\n", compartmentid, status);
+            return QUIC_STATUS_INVALID_PARAMETER;
+        } else {
+            WriteOutput("Running in Compartment %d\n", compartmentid);
+        }
+    }
+
+    uint8_t cpuCore;
+    if (TryGetValue(argc, argv, "core",  &cpuCore)) {
+        SetThreadAffinityMask(GetCurrentThread(), (DWORD_PTR)(1ull << cpuCore));
+    }
+#endif
+
+    TryGetValue(argc, argv, "sendbuf", &UseSendBuffer);
+
+    uint32_t ioSize = UseSendBuffer ? THROUGHPUT_DEFAULT_IO_SIZE_BUFFERED : THROUGHPUT_DEFAULT_IO_SIZE_NONBUFFERED;
+    TryGetValue(argc, argv, "iosize", &ioSize);
+    IoSize = ioSize;
+
+    uint32_t ioCount = UseSendBuffer ? THROGHTPUT_DEFAULT_SEND_COUNT_BUFFERED : THROUGHPUT_DEFAULT_SEND_COUNT_NONBUFFERED;
+    TryGetValue(argc, argv, "iocount", &ioCount);
+    IoCount = ioCount;
+
+    size_t Len = strlen(Target);
+    TargetData.reset(new char[Len + 1]);
+    QuicCopyMemory(TargetData.get(), Target, Len);
+    TargetData[Len] = '\0';
+
+    RawIoBuffer = new uint8_t[IoSize];
```

Comment:
Make sure to validate all allocations.

Reply:
Already fixed. Also disabling `throwing new` is an internal compiler flag. Because of how construction works, if new is ever used and fails to allocate, but doesn't throw, the constructor will AV. Unless you do what I do during the pool allocator and placement new.